src/mlir/dialect/VeronaOps.td

```diff
@@ -244,4 +245,117 @@ def Verona_FieldOp : Verona_Op<"field", [InClassScope]> {
     let verifier = [{ return success(); }];
 }
 
+def Verona_ContinueOp : Verona_Op<"continue", [Terminator, InLoopScope]> {
+    let summary = "continues any loop block";
+
+    let arguments = (ins);
+    let results = (outs);
+    let assemblyFormat = [{ attr-dict }];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_BreakOp : Verona_Op<"break", [Terminator, InLoopScope]> {
+    let summary = "breaks any loop region, returning control to previous region";
+
+    let arguments = (ins);
+    let results = (outs);
+    let assemblyFormat = [{ attr-dict }];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_LoopExitOp : Verona_Op<"loop_exit", [InLoopScope]> {
+    let summary = "breaks any loop block if condition is `false`";
+
+    let builders = [
+      OpBuilder<"OpBuilder &builder, OperationState &state, bool cond">
+    ];
+    let arguments = (ins BooleanType:$cond);
+    let results = (outs);
+    let assemblyFormat = [{ $cond `:` type($cond) attr-dict }];
+    let verifier = [{ return success(); }];
+}
+
+def Verona_LoopReturnOp : Verona_Op<"loop_return", [Terminator, InLoopScope]> {
```

Comment:
What I had in mind was a single `Verona_ReturnOp` operation, which can be used either in loops or directly in function. Rather than `InLoopScope`, it would have trait `HasParent<"FuncOp", "WhileOp">`. The AST lowering would only ever emit `verona.return`, never `return`.

Additionally you need a verifier that checks that the argument type matches the function's return type (just like the standard `return` does). Since my PR #206 already does that (as a verifier), I'm happy for you to just leave a TODO.

Reply:
I know but that approach has several drawbacks.

First, there are many types of returns. Return from function, return from regions (into their parent region) and return from function inside a region (which is what `loop_return` does). Both function return and region return work as expected even with Verona types, so we only need to add the third type.

Second, once we add free variable analysis (args, rets) and track all region captures and returns, we could just return to the encompassing region with a flag (saying immediate return) but that would be hard if the loop could function-return either of multiple values.

Third, eventually, loops will be lowered to CFG blocks, in which case, `verona.loop_return` is identical to just `return` and can be converted without any additional semantic checks. We get that for free. So if we complicate the lowering now, we'll complicate the lowering later. If we keep it simple now, we'll get it for free.

Fourth, I'm still not convinced we need a special return for sub-typing. If we do, then we'll need a special "everything", and that won't scale. As we discussed earlier, as reasonable implementation is to add casts until the types are known, and even keep the casts for generics up until the final lowering. The semantics is complete and the pattern is simple.

Finally, my own experience lowering different kinds of IR, including at least two other MLIR projects, is that the more special your dialect is, the harder it is to work with because you need to imbue special meaning to everything. The worst thing to make special is control flow, because that knowledge is often embedded into the language itself, with the assumption that no one in their right minds would re-implement branches and returns.

I'd ask that we keep it like that, for now at least, until we can have a proper conversation about sub-typing and its effects in the dialect.

Reply:
Let's leave this discussion for the next step, when we discuss functions and returns, if they have to be special or not.